include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -1246,6 +1301,273 @@ TensorProductMatrixSymmetricSumCollection<dim, Number, n_rows_1d>::
+        FullMatrix<Number> K(n_dofs_1D_with_overlap, n_dofs_1D_with_overlap);
+
+        // inner cell
+        for (unsigned int i = 0; i < n_dofs_1D; ++i)
+          for (unsigned int j = 0; j < n_dofs_1D; ++j)
+            {
+              const unsigned int i0 = i + n_overlap - 1;
+              const unsigned int j0 = j + n_overlap - 1;
+              M[i0][j0]             = M_ref[i][j] * cell_extend[d][1];
+              K[i0][j0]             = K_ref[i][j] / cell_extend[d][1];
+            }
+
+        // left neighbor or left boundary
+        if (bids[d][0] == LaplaceBoundaryType::internal_boundary)
+          {
+            // left neighbor
+            Assert(cell_extend[d][0] > 0.0, ExcInternalError());
+
+            for (unsigned int i = 0; i < n_overlap; ++i)
+              for (unsigned int j = 0; j < n_overlap; ++j)
+                {
+                  const unsigned int i0 = n_dofs_1D - n_overlap + i;
+                  const unsigned int j0 = n_dofs_1D - n_overlap + j;
+                  M[i][j] += M_ref[i0][j0] * cell_extend[d][0];
+                  K[i][j] += K_ref[i0][j0] / cell_extend[d][0];
+                }
+          }
+        else
+          {
+            if (bids[d][0] == LaplaceBoundaryType::dirichlet)
+              {
+                // left DBC
+                const unsigned i0 = n_overlap - 1;
+                clear_row_and_column(i0, M);
+                clear_row_and_column(i0, K);
+              }
+            else if (bids[d][0] == LaplaceBoundaryType::neumann)
+              {
+                // left NBC -> nothing to do
+              }
+            else
+              {
+                AssertThrow(false, ExcNotImplemented());
+              }
+          }
+
+        // reight neighbor or right boundary
+        if (bids[d][1] == LaplaceBoundaryType::internal_boundary)
+          {
+            Assert(cell_extend[d][2] > 0.0, ExcInternalError());
+
+            for (unsigned int i = 0; i < n_overlap; ++i)
+              for (unsigned int j = 0; j < n_overlap; ++j)
+                {
+                  const unsigned int i0 = n_overlap + n_dofs_1D + i - 2;
+                  const unsigned int j0 = n_overlap + n_dofs_1D + j - 2;
+                  M[i0][j0] += M_ref[i][j] * cell_extend[d][2];
+                  K[i0][j0] += K_ref[i][j] / cell_extend[d][2];
+                }
+          }
+        else
+          {
+            if (bids[d][1] == LaplaceBoundaryType::dirichlet)
+              {
+                // right DBC
+                const unsigned i0 = n_overlap + n_dofs_1D - 2;
+                clear_row_and_column(i0, M);
+                clear_row_and_column(i0, K);
+              }
+            else if (bids[d][1] == LaplaceBoundaryType::neumann)
+              {
+                // right NBC -> nothing to do
+              }
+            else
+              {
+                AssertThrow(false, ExcNotImplemented());
+              }
+          }
+
+        Ms[d] = M;
+        Ks[d] = K;
```

Comment:
Would it make sense to create a reference to `Ms[d]` and `Ks[d]` rather that copying here in the end?